src/example/SampleCustomLoader/sampleCustLoader.cpp

```diff
@@ -0,0 +1,436 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+#include <chrono>
+#include <condition_variable>
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <mutex>
+#include <string>
+#include <thread>
+#include <vector>
+
+#include <assert.h>
+#include <rapidjson/document.h>
+#include <spdlog/spdlog.h>
+
+#include "../../customloaderinterface.hpp"
+
+using namespace rapidjson;
+using namespace ovms;
+
+#define PATH_SIZE 10
+#define RSIZE_MAX_STR 4096
+
+#define SAMPLE_LOADER_OK 0
+#define SAMPLE_LOADER_ERROR 0x10
+
+#define SAMPLE_LOADER_IR_MODEL 0
+#define SAMPLE_LOADER_ONNX_MODEL 1
+#define SAMPLE_LOADER_BLOB_MODEL 2
+
+// Time in seconds at which model status will be checked
+#define MODEL_CHECK_PERIOD 10
+typedef std::pair<std::string, int> model_id_t;
+
+/* 
+ * This class implements am example custom model loader for OVMS.
+ * It derives the implementation from base class CustomLoaderInterface
+ * defined in ovms. The purpose this example is to demonistrate the 
+ * usage of various APIs defined in base class, parse loader specic
+ * parameters from the config file. 
+ *
+ * It reads the model files and returns the buffers to be loaded by the 
+ * model server.  
+ *
+ * Also, based on the contents on <model>.status file, it black lists the model
+ * or removes the model from blacklisting. During the periodic check on model 
+ * loader will unload/reload model based on blacklist.
+ */
+
+class custSampleLoader : public CustomLoaderInterface {
+private:
+    std::vector<model_id_t> models_loaded;
+    std::vector<model_id_t> models_watched;
+    std::map<model_id_t, std::string> models_enable;
+    std::map<model_id_t, bool> models_blacklist;
+    std::mutex map_mutex;
+    std::mutex models_watched_mutex;
+
+protected:
+    int extract_input_params(const std::string& basePath, int version, const std::string& loaderOptions,
+        std::string& binFile, std::string& modelFile, std::string& enableFile, int& modelType);
+
+    int load_files(std::string& binFile, std::string& modelFile, int modelType, char** modelBuffer,
+        char** binBuffer, int* modelLen, int* binLen);
+
+    // Variables needed to manage the periodic thread.
+    std::mutex cv_m;
+    std::condition_variable cv;
+    int watchIntervalSec = 0;
+    bool watcherStarted = false;
+    std::thread watcher_thread;
+
+public:
+    custSampleLoader();
+    ~custSampleLoader();
+
+    // Virtual functions of the base class defined here
+    CustomLoaderStatus loaderInit(const std::string& loader_path);
+    CustomLoaderStatus loaderDeInit();
+    CustomLoaderStatus unloadModel(const std::string& modelName, int version);
+    CustomLoaderStatus loadModel(const std::string& modelName, const std::string& basePath, const int version,
+        const std::string& loaderOptions, char** xmlBuffer, int* xmlLen, char** binBuffer,

```

Comment:
Why can't we just use unique_ptr instead of char**? The ownership of a buffer is clear, we get safety of clearing always event if exceptions happen.

Reply:
In the current implementation of example loader, the file is loaded in C++. However there is a possibility that we will be loading using C API which returns char*.  I am not sure whether I can create unique_ptr  and return from a function

Reply:
You can pass std::unique_ptr<T>& into function and then set the pointer in that function.

In case we really want to support C loaders then yes you cannot use unique_ptr. However you should wrap the raw pointer in OVMS as quickly as you get it and you are sure it was allocated into smart pointer to defend ourselves from unpredicted code execution path (such as exceptions).

That being said - is there requirement to support C loaders?

Reply:
Changed the code to use vector by reference and got rid of pointers